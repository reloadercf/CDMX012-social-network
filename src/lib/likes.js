import {
  getAuth, updateDoc, getFirestore, doc,
} from '../firebase-imports.js';
import { app } from './firebase-config.js';

const auth = getAuth(app);
export const currentUser = () => auth.currentUser;

// el usuario quiere dar like
export async function giveLike(userId, postId, likeList) {
  const db = getFirestore();
  const modifylikeList = likeList;
  modifylikeList.push(userId);
  const postRef = doc(db, 'posts', postId);
  await updateDoc(postRef, {
    likes: modifylikeList,
  });
}

// el usuario va a quitar su like
export async function dislike(userId, postId, likeList) {
  const db = getFirestore();
  const modifylikeList = likeList;
  modifylikeList.pop(userId);
  const postRef = doc(db, 'posts', postId);
  await updateDoc(postRef, {
    likes: modifylikeList,
  });
}
