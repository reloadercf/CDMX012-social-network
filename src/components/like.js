/* eslint-disable max-len */
import { dislike, giveLike } from '../lib/likes.js';

export const like = (click, likeCount, userId, postId, likeList) => { // (leDiLike?, #likesDePost, uidUsuarioActual, idPost, arrayDeLikes)
  const likeElement = document.createElement('div');
  likeElement.setAttribute('class', 'like-btn');
  likeElement.setAttribute('id', 'like-btn');
  if (click === true) {
    likeElement.innerHTML = '<img src= "./assets/heart.png" alt="like fill button">';
  } else {
    likeElement.innerHTML = '<img src= "./assets/like.png" alt="like empty button">';
  }
  const count = document.createElement('span');
  count.setAttribute('class', 'likeCount');
  count.textContent = likeCount;

  likeElement.append(count);
  let clicked = click;

  likeElement.addEventListener('click', (e) => { // Asignar un eventhandler al btn like
    e.preventDefault();
    if (!clicked) {
      clicked = true; // cambia status del click
      giveLike(userId, postId, likeList); // actualizar lista de likes
      likeElement.innerHTML = '<img src= "./assets/heart.png" alt="like fill button">';
      count.textContent = parseInt(count.textContent, 10) + 1;
      likeElement.append(count);
    } else {
      dislike(userId, postId, likeList);
      clicked = false;
      likeElement.innerHTML = '<img src= "./assets/like.png" alt="like empty button">';
      count.textContent = parseInt(count.textContent, 10) - 1; // actualizar lista de likes
      likeElement.append(count);
    }
  });
  return likeElement;
};
