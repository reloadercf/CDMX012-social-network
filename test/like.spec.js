/**
 * @jest-environment jsdom
 */
import { like } from '../src/components/like.js';
import { giveLike, dislike } from '../src/lib/likes.js';

jest.mock('../src/firebase-imports.js');

describe('Tests of like', () => {
  test('render like', () => {
    const renderlike = like();
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    rootDiv.appendChild(renderlike);
    expect(rootDiv.innerHTML).toMatchSnapshot();
  });
  test('cuando click es true', () => {
    const elements = [true, 2, 'saraiDos', 'sarai12', ['mario', 'charlotte']];
    const div = like(...elements);
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    rootDiv.appendChild(div);
    expect(rootDiv.innerHTML).toMatchSnapshot();
  });
  test('cuando click es false', () => {
    const elements = [false, 2, 'saraiDos', 'sarai12', ['mario', 'charlotte']];
    const div = like(...elements);
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    rootDiv.appendChild(div);
    expect(rootDiv.innerHTML.includes('./assets/like.png')).toBe(true);
  });
});

describe('the function giveLike modify the list of likes', () => {
  test('render a modified list', () => {
    const elements = ['charles', 'charles2022', ['maria', 'benny']];
    giveLike(...elements);
  });
});

describe('the function dislike modify the list of likes', () => {
  test('render a modified list', () => {
    const elements = ['charles', 'charles2022', ['maria', 'benny']];
    dislike(...elements);
  });
});

describe('conditional on click to be true', () => {
  test('render a modified list', () => {
    const elements = [false, 2, 'saraiDos', 'sarai12', ['mario', 'charlotte']];
    const div = like(...elements);
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    rootDiv.appendChild(div);
    const btn = document.querySelector('#like-btn');
    btn.dispatchEvent(new Event('click'));
    expect(rootDiv.innerHTML.includes('./assets/heart.png')).toBe(true);
  });
});

describe('conditional on click also to be true', () => {
  test('render a modified list', () => {
    const elements = [true, 2, 'saraiDos', 'sarai12', ['mario', 'charlotte']];
    const div = like(...elements);
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    rootDiv.appendChild(div);
    const btn = document.querySelector('#like-btn');
    btn.dispatchEvent(new Event('click'));
    expect(rootDiv.innerHTML.includes('./assets/like.png')).toBe(true);
  });
});
