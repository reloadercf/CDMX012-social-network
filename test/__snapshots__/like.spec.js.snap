// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of like cuando click es true 1`] = `"<div class=\\"like-btn\\" id=\\"like-btn\\"><img src=\\"./assets/heart.png\\" alt=\\"like fill button\\"><span class=\\"likeCount\\">2</span></div>"`;

exports[`Tests of like render like 1`] = `"<div class=\\"like-btn\\" id=\\"like-btn\\"><img src=\\"./assets/like.png\\" alt=\\"like empty button\\"><span class=\\"likeCount\\"></span></div>"`;
